src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs

```diff
@@ -342,14 +349,13 @@ public virtual void GenerateAnonymousFunctionConversionError(DiagnosticBag diagn
         // Returns the inferred return type, or null if none can be inferred.
         public BoundLambda Bind(NamedTypeSymbol delegateType)
         {
-            BoundLambda result;
-            if (!_bindingCache.TryGetValue(delegateType, out result))
+            if (_bindingCache.TryGetValue(delegateType, out var result))
             {
```

Comment:
Concurrency on these caches will be quite rare, and encountered solely based on concurrent calls into the compiler APIs. I think we are likely to be better served using a simple Dictionary with an explicit lock around its uses.